File: monitor.py
#!/usr/bin/env python3
import os
import sys
import time
import random
import signal
import sqlite3
import threading
import requests

from http.server import BaseHTTPRequestHandler, HTTPServer
from playwright.sync_api import sync_playwright, TimeoutError as PlaywrightTimeout

# ---- Health-check HTTP server ----
class HealthHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path in ("/", "/health"):
            self.send_response(200)
            self.end_headers()
            self.wfile.write(b"OK")
        else:
            self.send_response(404)
            self.end_headers()

def start_health_server(port: int = 10000):
    srv = HTTPServer(("", port), HealthHandler)
    threading.Thread(target=srv.serve_forever, daemon=True).start()
    return srv

# ---- Pushover alert ----
PUSH_APP_TOKEN = os.environ.get("PUSHOVER_APP_TOKEN") or os.environ["PUSHOVER_API_TOKEN"]
PUSH_USER_KEY  = os.environ["PUSHOVER_USER_KEY"]

def send_pushover(message: str):
    resp = requests.post(
        "https://api.pushover.net/1/messages.json",
        data={"token": PUSH_APP_TOKEN, "user": PUSH_USER_KEY, "message": message},
        timeout=10,
    )
    resp.raise_for_status()

# ---- SQLite persistence ----
DB_PATH = "state.db"
def init_db():
    conn = sqlite3.connect(DB_PATH, timeout=30, check_same_thread=False)
    conn.execute("""
      CREATE TABLE IF NOT EXISTS stock_state (
        url        TEXT PRIMARY KEY,
        last_state TEXT NOT NULL,
        updated_at TIMESTAMP NOT NULL
      );
    """)
    conn.commit()
    return conn

# ---- Stock check via productDetails XHR ----
def check_stock(page, url: str) -> str:
    # Expect the productDetails XHR while navigating
    with page.expect_response(
        lambda r: "productDetails" in r.url and "spuId=" in r.url,
        timeout=10_000
    ) as resp_info:
        page.goto(url, wait_until="networkidle", timeout=30_000)
    resp = resp_info.value

    try:
        data = resp.json()
        skus = data.get("data", {}).get("skus", [])
    except Exception:
        return "out"

    for sku in skus:
        if sku.get("stock", {}).get("onlineStock", 0) > 0:
            return "in"
    return "out"

def sleep_until_top_of_minute():
    now = time.time()
    delay = 60 - (now % 60)
    time.sleep(delay + random.uniform(0, 1))

# ---- Main service ----
def main():
    urls = [
        u.strip()
        for u in os.environ.get("PRODUCT_URLS", "").split(",")
        if u.strip()
    ]
    if not urls:
        print("No PRODUCT_URLS configured; exiting.")
        sys.exit(1)

    conn = init_db()
    cur = conn.cursor()
    health_srv = start_health_server()

    def clean_exit(*_):
        print("Shutting down...")
        try: conn.close()
        except: pass
        try: health_srv.shutdown()
        except: pass
        sys.exit(0)

    signal.signal(signal.SIGINT, clean_exit)
    signal.signal(signal.SIGTERM, clean_exit)

    print("Launching browser…")
    with sync_playwright() as p:
        browser = p.chromium.launch(
            headless=True,
            args=[
                "--no-sandbox",
                "--disable-dev-shm-usage",
                "--lang=en-US",
                "--window-size=1920,1080",
            ]
        )
        context = browser.new_context(
            user_agent=(
                f"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                f"AppleWebKit/537.36 (KHTML, like Gecko) "
                f"Chrome/{random.randint(100,116)}.0.0.0 Safari/537.36"
            ),
            viewport={"width": 1920, "height": 1080},
            locale="en-US",
        )
        # Block image requests for speed
        context.route("**/*.{png,jpg,jpeg,svg,webp,gif}", lambda route: route.abort())
        page = context.new_page()

        print("Monitor started. Polling every minute.")
        while True:
            sleep_until_top_of_minute()
            for url in urls:
                state = None
                for attempt in range(1, 4):
                    try:
                        state = check_stock(page, url)
                        break
                    except PlaywrightTimeout as e:
                        print(f"[Attempt {attempt}] timeout on {url}: {e}", file=sys.stderr)
                    except Exception as e:
                        print(f"[Attempt {attempt}] error on {url}: {e}", file=sys.stderr)
                        time.sleep(attempt)
                if state is None:
                    print(f"All attempts failed for {url}", file=sys.stderr)
                    continue

                cur.execute("SELECT last_state FROM stock_state WHERE url = ?", (url,))
                row = cur.fetchone()
                old = row[0] if row else "out"
                if old == "out" and state == "in":
                    print(f"{time.strftime('%Y-%m-%d %H:%M:%S')} {url} → IN STOCK!")
                    try:
                        send_pushover(f"🔥 In stock: {url}")
                    except Exception as e:
                        print(f"Pushover error: {e}", file=sys.stderr)

                if not row:
                    cur.execute(
                        "INSERT INTO stock_state(url, last_state, updated_at) "
                        "VALUES (?, ?, CURRENT_TIMESTAMP)",
                        (url, state)
                    )
                else:
                    cur.execute(
                        "UPDATE stock_state "
                        "SET last_state = ?, updated_at = CURRENT_TIMESTAMP "
                        "WHERE url = ?",
                        (state, url)
                    )
                conn.commit()

        browser.close()

if __name__ == "__main__":
    main()

File: monitor.js
// monitor.js

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

// ─── Configuration ────────────────────────────────────────────────────────────

// Comma-separated list of Popmart product URLs
const URLs = (process.env.POP_MART_URLS || '')
  .split(',')
  .map(u => u.trim())
  .filter(Boolean);

// Pushover credentials
const TOKEN = process.env.PUSHOVER_TOKEN;
const USER  = process.env.PUSHOVER_USER_KEY;

// Timeouts (ms)—defaults used in testing:
const NAV_TIMEOUT      = parseInt(process.env.NAV_TIMEOUT, 10)      || 30000; // 30 s
const SELECTOR_TIMEOUT = parseInt(process.env.SELECTOR_TIMEOUT, 10) || 15000; // 15 s
const RENDER_WAIT      = parseInt(process.env.RENDER_WAIT, 10)      || 2000;  // 2 s

// State file to track which URLs have already alerted
const STATE_FILE = path.resolve(__dirname, 'state.json');

// ─── Sanity checks ─────────────────────────────────────────────────────────────
if (!URLs.length) {
  console.error('❌ POP_MART_URLS must be set');
  process.exit(1);
}
if (!TOKEN || !USER) {
  console.error('❌ PUSHOVER_TOKEN and PUSHOVER_USER_KEY must be set');
  process.exit(1);
}

// ─── Load or initialize state ─────────────────────────────────────────────────
let state = {};
if (fs.existsSync(STATE_FILE)) {
  try {
    state = JSON.parse(fs.readFileSync(STATE_FILE, 'utf8'));
  } catch {
    console.warn('⚠ Could not parse state.json; starting fresh.');
  }
}
for (const url of URLs) {
  if (typeof state[url] !== 'boolean') {
    state[url] = false;
  }
}

// ─── Helper: simple retry wrapper ───────────────────────────────────────────────
async function withRetries(fn, retries = 3, delay = 500) {
  let attempt = 0;
  while (true) {
    try {
      return await fn();
    } catch (err) {
      attempt++;
      console.error(`   ❌ Attempt ${attempt} failed: ${err.message}`);
      if (attempt >= retries) throw err;
      console.log(`   ⏳ Retrying in ${delay}ms…`);
      await new Promise(r => setTimeout(r, delay));
    }
  }
}

// ─── Main ──────────────────────────────────────────────────────────────────────
(async () => {
  console.log('▶️  Starting monitor');

  // 1) Launch a single browser + context + page
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    userAgent:
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 13_5) ' +
      'AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.4 Safari/605.1.15',
    viewport: { width: 1280, height: 800 },
  });
  const page = await context.newPage();

  // 2) Accept T&C modal once (on initial load)
  try {
    await page.goto(URLs[0], { waitUntil: 'domcontentloaded', timeout: NAV_TIMEOUT });
    const acceptBtn = page.locator('button', { hasText: /^accept$/i }).first();
    if (await acceptBtn.count()) {
      console.log('    ▶ Clicking initial Accept modal');
      await acceptBtn.click({ timeout: 5000 });
      await page.waitForTimeout(1000);
    }
  } catch (e) {
    console.warn('⚠ Initial T&C acceptance step failed:', e.message);
  }

  // 3) Loop through each URL
  for (const url of URLs) {
    const wasInStock = state[url];
    console.log(`\n🔗  Checking ${url}  (previously inStock: ${wasInStock})`);

    try {
      const nowInStock = await withRetries(async () => {
        // 3a) Navigate
        try {
          await page.goto(url, { waitUntil: 'domcontentloaded', timeout: NAV_TIMEOUT });
        } catch {
          console.warn('    ⚠ domcontentloaded timed out; using simple goto');
          await page.goto(url, { timeout: NAV_TIMEOUT });
        }

        // 3b) Wait for React to render dynamic button
        await page.waitForTimeout(RENDER_WAIT);

        // 3c) Wait for the stock-button container
        await page.waitForSelector(
          'div[class*="index_usBtn__"][class*="index_btnFull__"]',
          { timeout: SELECTOR_TIMEOUT }
        );

        // 3d) Detect “ADD TO BAG”
        const addCount = await page.locator(
          'div[class*="index_usBtn__"][class*="index_btnFull__"]:has-text("ADD TO BAG")'
        ).count();

        return addCount > 0;
      });

      console.log(`    ✓ nowInStock = ${nowInStock}`);

      // 4) Notify only on a fresh restock
      if (!wasInStock && nowInStock) {
        console.log('    ▶ New restock! Sending Pushover alert');
        const title = await page.title();
        await fetch('https://api.pushover.net/1/messages.json', {
          method: 'POST',
          body: new URLSearchParams({
            token: TOKEN,
            user: USER,
            message: `${title} is back in stock → ${url}`,
          }),
        });
        console.log('    ✓ Alert sent');
      } else {
        console.log('    • No alert (still out or already notified)');
      }

      // Update state for this URL
      state[url] = nowInStock;
    } catch (err) {
      console.error(`Error checking ${url}:`, err.message);
    }
  }

  // 5) Persist updated state
  try {
    fs.writeFileSync(STATE_FILE, JSON.stringify(state, null, 2), 'utf8');
    console.log('\n✅ State saved to state.json');
  } catch (e) {
    console.error('❌ Failed to save state.json:', e.message);
  }

  // 6) Teardown
  await browser.close();
  console.log('\n🏁 All URLs processed');
})();
